### PR TITLE
(FACT-2330) Add ssh fact for Windows OpenSSH feature

### PR DESCRIPTION
**Pull request by oanatmaria**

### CODE REVIEW

File: lib/resolvers/ssh_resolver.rb [hunk 1]
```diff
@@ -24,52 +21,16 @@ def retrieve_info(fact_name)
             next unless File.directory?(file_path)
 
             @file_names.each do |file_name|
-              next unless FileTest.file?(File.join(file_path, file_name))
+              next unless File.readable?(File.join(file_path, file_name))
 
               key_type, key = File.read(File.join(file_path, file_name)).split(' ')
-              key_name = determine_ssh_key_name(key_type)
-              ssh_list << create_ssh(key_name, key_type, key)
+              key_name = SshHelper.determine_ssh_key_name(key_type)
```
Comment (sebastian-miclea) [inline lib/resolvers/ssh_resolver.rb:27]: Could we do this in the SshHelper? I don't see key_name being used anywhere else besides calling an other method from SshHelper.

File: lib/resolvers/utils/ssh_helper.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        ssh_fp = SSH_FINGERPRINT[key_name]
+        sha1 = "SSHFP #{ssh_fp} 1 #{Digest::SHA1.new.update(decoded_key)}"
+        sha256 = "SSHFP #{ssh_fp} 2 #{Digest::SHA2.new.update(decoded_key)}"
+
+        fingerprint = FingerPrint.new(sha1, sha256)
+        Ssh.new(fingerprint, key_type, key, key_name)
+      end
+
+      def determine_ssh_key_name(key)
+        SSH_NAME[key]
```
Comment (sebastian-miclea) [inline lib/resolvers/utils/ssh_helper.rb:24]: I don't think that a method is needed to access a key from a hash.

File: lib/framework/core/file_loader.rb [hunk 3]
```diff
@@ -32,6 +32,7 @@ def load_lib_dirs(*dirs)
 require "#{ROOT_DIR}/lib/framework/config/block_list"
 require "#{ROOT_DIR}/lib/resolvers/utils/fingerprint.rb"
 require "#{ROOT_DIR}/lib/resolvers/utils/ssh.rb"
+require "#{ROOT_DIR}/lib/resolvers/utils/ssh_helper.rb"
```
Comment (gimmyxd) [inline lib/framework/core/file_loader.rb:35]: Why there is a util/shh and a util/ssh_helper?
This is a bit confusing, alongside the fact that utils are in different locations with different purposes. Could this 2 ssh be a single source?

File: lib/facts/windows/ssh.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+module Facts
+  module Windows
+    class Ssh
+      FACT_NAME = 'ssh'
+
+      def call_the_resolver
+        privileged = Facter::Resolvers::Identity.resolve(:privileged)
+        result = Facter::Resolvers::Windows::Ssh.resolve(:ssh) if privileged
```
Comment (IrimieBogdan) [inline lib/facts/windows/ssh.rb:10]: `result` variable should have a suggestive name

File: lib/resolvers/ssh_resolver.rb (hunk 1 above)
Comment (oanatmaria) [reply lib/resolvers/ssh_resolver.rb]: Done

File: lib/resolvers/utils/ssh_helper.rb (hunk 2 above)
Comment (oanatmaria) [reply lib/resolvers/utils/ssh_helper.rb]: I removed it

File: lib/facts/windows/ssh.rb (hunk 4 above)
Comment (oanatmaria) [reply lib/facts/windows/ssh.rb]: done

File: lib/framework/core/file_loader.rb (hunk 3 above)
Comment (oanatmaria) [reply lib/framework/core/file_loader.rb:35]: util/shh is an object that contains structured information about a ssh key. util/ssh_helper creates an object of type util/shh after creating a fingerprint and extracting other information

File: lib/resolvers/utils/ssh_helper.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+require 'base64'
+require 'digest/sha1'
+
+module Facter
+  class SshHelper
```
Comment (Filipovici-Andrei) [inline lib/resolvers/utils/ssh_helper.rb:7]: Shouldn't you also have a module named Utils?

File: spec/facter/facts/windows/ssh_spec.rb [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+      it 'calls Facter::Resolvers::Windows::Identity' do
+        fact.call_the_resolver
+        expect(Facter::Resolvers::Identity).to have_received(:resolve).with(:privileged)
+      end
+
+      it 'returns ssh fact' do
+        expect(fact.call_the_resolver).to be_an_instance_of(Facter::ResolvedFact).and \
+          have_attributes(name: 'ssh', value: value)
+      end
```
Comment (Filipovici-Andrei) [inline spec/facter/facts/windows/ssh_spec.rb:84]: This two tests can be put into a shared example

File: spec/facter/resolvers/windows/ssh_spec.rb [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+          expect(Facter::Resolvers::Windows::Ssh.resolve(:ssh)).to eq([rsa_result, ecdsa_result, ed25519_result])
+        end
+      end
+
+      context 'when files are not readable' do
+        let(:ecdsa_exists) { false }
+        let(:rsa_exists) { false }
+        let(:ed25519_exists) { false }
+
+        it 'returns nil' do
```
Comment (Filipovici-Andrei) [inline spec/facter/resolvers/windows/ssh_spec.rb:89]: Probably also a test when just one file is not readable, might be a good test

**Review by Filipovici-Andrei — CHANGES_REQUESTED** [review]

File: spec/facter/resolvers/windows/ssh_spec.rb (hunk 7 above)
Comment (oanatmaria) [reply spec/facter/resolvers/windows/ssh_spec.rb:89]: in context 'when ecdsa, ed25519 and rsa files exists' do implays that dsa file is not readable

File: lib/resolvers/utils/ssh_helper.rb (hunk 5 above)
Comment (oanatmaria) [reply lib/resolvers/utils/ssh_helper.rb:7]: Done

File: spec/facter/facts/windows/ssh_spec.rb (hunk 6 above)
Comment (oanatmaria) [reply spec/facter/facts/windows/ssh_spec.rb:84]: We decided that a shared example will degenerate over time so we will avoid those

File: spec/facter/resolvers/windows/ssh_spec.rb (hunk 7 above)
Comment (Filipovici-Andrei) [reply spec/facter/resolvers/windows/ssh_spec.rb:89]: Ok

File: spec/facter/facts/windows/ssh_spec.rb (hunk 6 above)
Comment (Filipovici-Andrei) [reply spec/facter/facts/windows/ssh_spec.rb:84]: Anything can degenerate if we do stupid things. And that shouldn't be a reason to completely avoid using a programming paradigm.

File: lib/resolvers/utils/ssh_helper.rb (hunk 5 above)
Comment (Filipovici-Andrei) [reply lib/resolvers/utils/ssh_helper.rb:7]: I think the correct namespace should be Facter::Resolvers::Utils::SshHelper

File: lib/resolvers/windows/ssh.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+          end
+
+          def retrieve_info(fact_name)
+            ssh_dir = determine_ssh_dir
+            return unless ssh_dir && File.directory?(ssh_dir)
+
+            ssh_list = []
+
+            FILE_NAMES.each do |file_name|
+              next unless File.readable?(File.join(ssh_dir, file_name))
```
Comment (Filipovici-Andrei) [inline lib/resolvers/windows/ssh.rb:26]: You should use the FileHelper.safe_read

**Review by Filipovici-Andrei — CHANGES_REQUESTED** [review]

File: lib/resolvers/windows/ssh.rb (hunk 8 above)
Comment (oanatmaria) [reply lib/resolvers/windows/ssh.rb]: Done

**Review by Filipovici-Andrei — APPROVED** [review]

File: lib/facts/windows/ssh.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+        privileged = Facter::Resolvers::Identity.resolve(:privileged)
+        ssh_info = Facter::Resolvers::Windows::Ssh.resolve(:ssh) if privileged
+        ssh_facts = {}
+        ssh_info&.each { |ssh| ssh_facts.merge!(create_ssh_fact(ssh)) }
+        Facter::ResolvedFact.new(FACT_NAME, ssh_facts.empty? ? nil : ssh_facts)
+      end
+
+      private
+
+      def create_ssh_fact(ssh)
```
Comment (gimmyxd) [inline lib/facts/windows/ssh.rb:18]: Isn't this duplicated in each ssh fact? is there a ticket to solve the duplications?

**Pull request merged**